### ISSUE TITLE
FreeBSD14 error(dco_update_peer_stat: invalid peer ID 2 returned by kernel)

### ISSUE THREAD

**Issue by githubrn** [OpenVPN/openvpn#487]
**Describe the bug**
1. OpenVPN send short buffer in **ioctl()**. 
2. FreeBSD sys/net/if_ovpn.c return **ENOSPC**, need: **ENOBUF**

in openvpn.log messange 

`2024-01-20 10:50:14 xxxxxx/aaaaa:45216 dco_update_peer_stat: invalid peer ID 2 returned by kernel`

**To Reproduce**
clean install from ports

**Expected behavior**
no errors in logs, work DCO.

**Version information (please complete the following information):**
```
root@-14:/ # pkg info -x openvpn
openvpn-2.6.8_1
```
```
root@-14:/ # uname -a
FreeBSD gw-de-14 14.0-RELEASE FreeBSD 14.0-RELEASE #0 releng/14.0-n265380-f9716eee8ab4: Fri Nov 10 05:57:23 UTC 2023     root@releng1.nyi.freebsd.org:/usr/obj/usr/src/amd64.amd64/sys/GENERIC amd64

```

**Additional context**
debug in freebsd:

```
root@-14:/ # cd /tmp/
root@-14:/tmp # fetch http://www.grosbein.net/freebsd/patches/if_ovpn.c.diff
if_ovpn.c.diff 1133 B 6916 kBps 00s
root@-14:/tmp # cd /usr/src/sys/net
​
root@-14:/usr/src/sys/net # patch < /tmp/if_ovpn.c.diff
Hmm... Looks like a unified diff to me...
The text leading up to this was:
--------------------------
|--- if_ovpn.c.orig 2023-11-10 07:07:53.000000000 +0700
|+++ if_ovpn.c 2024-01-19 22:30:53.478336000 +0700
--------------------------
Patching file if_ovpn.c using Plan A...
Hunk #1 succeeded at 48.
Hunk #2 succeeded at 277.
Hunk #3 succeeded at 1419.
done
root@-14:/usr/src/sys/net # cd /usr/src/sys/modules/if_ovpn
root@-14:/usr/src/sys/modules/if_ovpn # make
[Creating objdir /usr/obj/usr/src/amd64.amd64/sys/modules/if_ovpn...]
machine -> /usr/src/sys/amd64/include
......skip
root@-14:/usr/src/sys/modules/if_ovpn # service openvpn stop
Stopping openvpn.
Waiting for PIDS: 52148.
root@-14:/usr/src/sys/modules/if_ovpn # kldunload if_ovpn
root@-14:/usr/src/sys/modules/if_ovpn # kldload /usr/obj/usr/src/amd64.amd64/sys/modules/if_ovpn/if_ovpn.ko
root@-14:/usr/src/sys/modules/if_ovpn # service openvpn start
Starting openvpn.
root@-14:/usr/src/sys/modules/if_ovpn # sysctl net.link.openvpn.debug=1
net.link.openvpn.debug: 0 -> 1
```
_in /var/log/debug.log_

```
root@-14:/usr/local/etc/wireguard # tail -f /var/log/debug.log
Jan 22 17:29:26 -14 kernel: if_ovpn: len(4594) > ifd->ifd_len(4096)
Jan 22 17:30:06 -14 syslogd: last message repeated 1 times
Jan 22 17:32:07 -14 syslogd: last message repeated 2 times
Jan 22 17:35:07 -14 syslogd: last message repeated 3 times
Jan 22 17:35:59 -14 kernel: if_ovpn: len(4269) > ifd->ifd_len(4096)
Jan 22 17:36:07 -14 syslogd: last message repeated 1 times
Jan 22 17:38:07 -14 syslogd: last message repeated 2 times
Jan 22 17:41:07 -14 syslogd: last message repeated 3 times
Jan 22 18:16:21 -14 syslogd: last message repeated 1 times
Jan 22 18:21:25 -14 syslogd: last message repeated 5 times
```
_correction, patch src openvpn FreeBSD14 in /usr/ports:_

```
cd /usr/ports/security/openvpn/files
fetch http://www.grosbein.net/freebsd/patches/patch-src_openvpn_dco__freebsd.c
cd ..
make clean all deinstall install
service openvpn restart
```
[freebsd14_and_openvpn.zip](https://github.com/OpenVPN/openvpn/files/14035967/freebsd14_and_openvpn.zip)

**Comment by cron2** [comment]
@kprovost can you have a look?  I guess you're the local expert here, since both sides of this communication is your code :-)

**Comment by kprovost** [comment]
The reporter opened a bug in FreeBSD's bugzilla too: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=276583

To be honest, I can't work out what they're trying to say. I've asked for clarification in the bugzilla ticket.

**Comment by cron2** [comment]
My reading in all these lines is that OpenVPN userland is passing a buffer to kernel land which is "sometimes" too small for what the kernel wants to send back.  But I admit I did not read all of it, as it's lots of text that is not relevant to the problem report - including the actual diffs would be more helpful than describing how to apply the diff and rebuild stuff.

**Comment by kprovost** [comment]
Yeah, I got clarification on that in the bug report.
The real issue is indeed that the buffer in dco_get_peer_stats_multi() is too small. The rest is mostly irrelevant.

I'll take a look at the code later today to see if the bump to 8k is consistently enough or not.

On a semi-related (because that would avoid the entire buffer size problem) note: there may be some interest on the Netgate side to convert the if_ovpn driver to use netlink rather than ioctl/nvlist as it is today. Ideally using the same code as Linux. I hope I'm correct in assuming that that'd be welcome in OpenVPN.

**Comment by Pippin1st** [comment]
Just for info, here some more:
https://redmine.pfsense.org/issues/15154
https://forum.netgate.com/topic/185412/new-user-of-openvpn-on-pfsense-dco_update_peer_stat-invalid-peer-id-0-returned-by-kernel
https://forums.openvpn.net/viewtopic.php?t=43476

**Comment by kprovost** [comment]
> Just for info, here some more: https://redmine.pfsense.org/issues/15154 https://forum.netgate.com/topic/185412/new-user-of-openvpn-on-pfsense-dco_update_peer_stat-invalid-peer-id-0-returned-by-kernel https://forums.openvpn.net/viewtopic.php?t=43476

 None of those appear to be related to the buffer size issue though.

**Comment by ordex** [comment]
> Ideally using the same code as Linux. I hope I'm correct in assuming that that'd be welcome in OpenVPN.

That would be amazing, not just welcome! I'd love to see sitnl re-used for platforms other than Linux :)

**Comment by ordex** [comment]
> My reading in all these lines is that OpenVPN userland is passing a buffer to kernel land which is "sometimes" too small for what the kernel wants to send back. But I admit I did not read all of it, as it's lots of text that is not relevant to the problem report - including the actual diffs would be more helpful than describing how to apply the diff and rebuild stuff.

```
--- if_ovpn.c.orig	2023-11-10 07:07:53.000000000 +0700
+++ if_ovpn.c	2024-01-19 22:30:53.478336000 +0700
@@ -48,6 +48,7 @@
 #include <sys/socketvar.h>
 #include <sys/sockio.h>
 #include <sys/sysctl.h>
+#include <sys/syslog.h>
 #include <sys/time.h>
 
 #include <machine/atomic.h>
@@ -276,6 +277,12 @@ SYSCTL_INT(_net_link_openvpn, OID_AUTO, netisr_queue,
 	CTLFLAG_VNET | CTLFLAG_RW, &VNET_NAME(async_netisr_queue), 0,
 	"Use netisr_queue() rather than netisr_dispatch().");
 
+VNET_DEFINE_STATIC(int, debug) = 0;
+#define	V_debug		VNET(debug)
+SYSCTL_INT(_net_link_openvpn, OID_AUTO, debug,
+	CTLFLAG_VNET | CTLFLAG_RW, &VNET_NAME(debug), 0,
+	"Enable debug log");
+
 static int
 ovpn_peer_compare(struct ovpn_kpeer *a, struct ovpn_kpeer *b)
 {
@@ -1412,7 +1419,10 @@ ovpn_ioctl_get(struct ifnet *ifp, struct ifdrv *ifd)
 		if (len > ifd->ifd_len) {
 			free(packed, M_NVLIST);
 			nvlist_destroy(nvl);
-			return (ENOSPC);
+			if (V_debug)
+			  log(LOG_DEBUG, "if_ovpn: len(%ju) > ifd->ifd_len(%ju)\n",
+			    (uintmax_t)len, (uintmax_t)ifd->ifd_len);
+			return (ENOBUFS);
 		}
 
 		error = copyout(packed, ifd->ifd_data, len);

```

**Comment by cron2** [comment]
> On a semi-related (because that would avoid the entire buffer size problem) note: there may be some interest on the Netgate side to convert the if_ovpn driver to use netlink rather than ioctl/nvlist as it is today. Ideally using the same code as Linux. I hope I'm correct in assuming that that'd be welcome in OpenVPN.

I have no strong opinion on this.  The current FreeBSD code works and I understand what it does, so there is no immediate need to change it.  OTOH, if the netlink is similar enough to the Linux code and can actually share code parts, this would help future maintenance.  I'm not a netlink expert on Linux, so I have no idea how similar this would be... (and whether Linux DCO will actually stay on netlink, or move elsewhere - we should discuss this in a round that includes @schwabe and @ordex )

**Comment by ordex** [comment]
> whether Linux DCO will actually stay on netlink, or move elsewhere

There is no plan to move Linux DCO away from Netlink, as it is the currently preferred user<->kernel interface Linux wide.

**Comment by cron2** [comment]
> > Ideally using the same code as Linux. I hope I'm correct in assuming that that'd be welcome in OpenVPN.
> 
> That would be amazing, not just welcome! I'd love to see sitnl re-used for platforms other than Linux :)

I am not hearing "reuse of sitnl"... while that is a netlink-enabling API inside OpenVPN userland (for ifconfig/route), it's not "the thing linux DCO uses", no?  So converting dco_freebsd.c -> if_dco.c "to use netlink" would not mean "sitnl".

**Comment by kprovost** [comment]
> I am not hearing "reuse of sitnl"... while that is a netlink-enabling API inside OpenVPN userland (for ifconfig/route), it's not "the thing linux DCO uses", no? So converting dco_freebsd.c -> if_dco.c "to use netlink" would not mean "sitnl".

That's correct, yes. While there is some work in FreeBSD to use netlink for routing configuration as well I don't think that's at the point where openvpn could reasonably use it on both Linux and FreeBSD. The potential netlink change is just about DCO configuration. And there's some dealing with older versions (both new openvpn on freebsd 14 and old openvpn on freebsd 15) to figure out as well.

**Comment by kprovost** [comment]
I've posted a fix for the buffer size issue on the mailing list: https://sourceforge.net/p/openvpn/mailman/message/58728397/

**Comment by cron2** [comment]
Patch looks good.  Of course I'd like to test this :-) - so what do I need to do to trigger the problem?  "Have 1000 clients active at the same time"?

(To test the realloc() thing, starting with bufsize=16 should work, but I'd also like to see the original problem)

**Comment by kprovost** [comment]
Yeah, that's pretty much the only way.
The issue will manifest if the nvlist grows to be too large. It really only scales with number of peers. "Happily" nvlist serialisation isn't particularly space-efficient so it's not as if you're going to need 1000, but it may be a fair few. I'm completely guessing here: 64 to 128 peers.

**Comment by cron2** [comment]
Patch has been merged, and will be in the upcoming 2.6.9 release

commit 62676935d738f74908845ca96819a36a8c0c230e
Author: Kristof Provost <kp@FreeBSD.org>
Date:   Wed Jan 24 16:27:39 2024 +0100

    dco-freebsd: dynamically re-allocate buffer if it's too small

(and  d8faf568d237667c66141e2c3f6df3f999aa02bd for release/2.6)
